### PR TITLE
Refactor consult--*-builder

### PR DESCRIPTION
**Pull request by aagon**
Like discussed in #616, a small proof of concept regarding the refactorisation and customisation of exclude arguments for consult-{grep,ripgrep,find,git-grep}.

Posted here for review by @minad.

A short remark : the `consult--build-arguments` function is such that every function inside the `consult-grep-args` must generate a list of split strings. We might want make it a little more robust. Since the function is idempotent, a simple idea to support functions producing unsplit strings is to replace `(consult--build-arguments (consult-grep-args))` with `(consult--build-arguments (consult--build-arguments (consult-grep-args)))`.

I have left `consult--grep-generate-exclude-args-project` for reference and if users want to use it, but haven't put it in the default value of `consult-grep-args`. The existence of this ready-to-use function should be advertised somewhere.

If you agree with this implementation, I can apply it to ripgrep, find and git-grep.

### CODE REVIEW

**Comment by minad** [comment]
Thanks! Please adjust all the other commands too.

I don't want to call the splitter twice, even if the function is idempotent. I want to keep it simple.

I will add more comments to the code.

File: consult.el [hunk 1]
```diff
@@ -236,10 +236,12 @@ See `consult--multi' for a description of the source data structure."
   "Regexp used to match file and line of grep output.")
 
 (defcustom consult-grep-args
-  "grep --null --line-buffered --color=never --ignore-case --line-number -I -r ."
+  '("grep" (consult--grep-generate-exclude-args-common)
+    "--null" "--line-buffered" "--color=never"
+    "--ignore-case" "--line-number" "-I" "-r" ".")
```
Comment (minad) [inline consult.el:241]: You don't have to split all the strings here, since we us split-string-and-unquote.

File: consult.el [hunk 2]
```diff
@@ -4405,6 +4415,25 @@ FIND-FILE is the file open function, defaulting to `find-file'."
...
+  "Produce grep exclude arguments based on `grep-find-ignored-files'."
+  (unless (boundp 'grep-find-ignored-files) (require 'grep))
+  (append
+   (mapcar (lambda (s) (concat "--exclude=" s))
+	   (bound-and-true-p grep-find-ignored-files))
+   (mapcar (lambda (s) (concat "--exclude-dir=" s))
+	   (bound-and-true-p grep-find-ignored-directories))))
+
+(defun consult--grep-generate-exclude-args-project ()
+  "Produce grep exclude arguments based on `project-ignores'."
```
Comment (minad) [inline consult.el:4428]: I don't want to provide that function here. Users can add this code to their user configuration.

File: consult.el [hunk 3]
```diff
@@ -1106,6 +1108,14 @@ MARKER is the cursor position."
   (let ((inhibit-field-text-motion t))
     (consult--region-with-cursor (line-beginning-position) (line-end-position) marker)))
 
+(defun consult--build-arguments (str)
+  "Return arg as flat list of split strings."
```
Comment (minad) [inline consult.el:1112]: Please expand the docstring here.

Comment (minad) [reply consult.el:1112]: Do you have package-lint enabled? It should warn you about the undocumented argument.

File: consult.el (hunk 1 above)
Comment (aagon) [reply consult.el]: done

File: consult.el (hunk 2 above)
Comment (aagon) [reply consult.el]: removed. Shall I add it to the wiki ?

File: consult.el (hunk 3 above)
Comment (aagon) [reply consult.el]: Done, tell me what you think.

File: consult.el [hunk 4]
```diff
@@ -236,10 +236,12 @@ See `consult--multi' for a description of the source data structure."
...
 
 (defcustom consult-grep-args
-  "grep --null --line-buffered --color=never --ignore-case --line-number -I -r ."
+  '("grep" (consult--grep-generate-exclude-args-common)
+    "--null" "--line-buffered" "--color=never"
+    "--ignore-case" "--line-number" "-I" "-r" ".")
   "Command line arguments for grep, see `consult-grep'.
 The dynamically computed arguments are appended."
-  :type 'string)
+  :type '(choice string list))
```
[the remark this answers is not in the document]
Comment (aagon) [reply consult.el]: done

File: consult.el [hunk 5]
```diff
@@ -236,10 +236,12 @@ See `consult--multi' for a description of the source data structure."
   "Regexp used to match file and line of grep output.")
 
 (defcustom consult-grep-args
-  "grep --null --line-buffered --color=never --ignore-case --line-number -I -r ."
+  '("grep" (consult--grep-generate-exclude-args-common)
+    "--null --line-buffered --color=never\
+    --ignore-case --line-number -I -r .")
   "Command line arguments for grep, see `consult-grep'.
```
Comment (minad) [inline consult.el:242]: The docstring should be expanded. It should mention both possibilities - strings and lists with evaluated arguments.

**Comment by minad** [comment]
Looks good. Can you please also adjust the defcustoms (types and do strings) of the other commands? find, git-grep, ripgrep, man. Then it should be ready to merge. I can do this later.

**Comment by aagon** [comment]
I think I have applied all your comments. I have also adjusted consult-locate.

File: consult.el [hunk 6]
```diff
@@ -236,42 +236,56 @@ See `consult--multi' for a description of the source data structure."
...
+  '("grep" (consult--grep-generate-exclude-args)
+    "--null --line-buffered --color=never\
+    --ignore-case --line-number -I -r .")
   "Command line arguments for grep, see `consult-grep'.
-The dynamically computed arguments are appended."
-  :type 'string)
+The dynamically computed arguments are appended.
+
+Can be either a string, or a list of string or expressions."
+  :type '(choice string (repeat string)))
```
Comment (minad) [inline consult.el:246]: Sorry, what I mentioned about the type before was wrong. Should be (choice string (repeat (choice string expression))).

File: consult.el [hunk 7]
```diff
@@ -4405,6 +4419,15 @@ FIND-FILE is the file open function, defaulting to `find-file'."
       (substring cand (1+ (length (get-text-property 0 'consult--grep-file cand))))
     (get-text-property 0 'consult--grep-file cand)))
 
+(defun consult--grep-generate-exclude-args-common ()
```
[the remark this answers is not in the document]
Comment (aagon) [reply consult.el]: done

File: consult.el (hunk 5 above)
Comment (aagon) [reply consult.el:242]: done

**Comment by minad** [comment]
Thanks. Did you assign copyright to the FSF for your Emacs contributions? See https://github.com/minad/consult#contributions

**Comment by minad** [comment]
Merged in https://github.com/minad/consult/commit/fae2c56bdc6832af65b85a3f2111be32eb2e68b2

**Pull request closed without merging**